### PR TITLE
Add `version` field to ink! metadata

### LINKED ISSUE #1289: Update `MetadataVersioned` to have `version` field
Right now versioning in the ink! metadata is done through the use of a version key which
wraps the metadata object:

```
"V3": {
    ... ink! metadata object ...
}
```

We should instead change this to be better defined by introducing a seperate `version`
field into the ink! metadata.

```
"abi": {
  "version": "3",
  "spec": { ... },
  "storage": { ... },
  "types": { ... },
}
```

This makes the meaning of the field more clear. We'll want to introduce this along side
the `V4` metadata changes.

### PR DESCRIPTION
**Pull request by HCastano**
Up until now the version of the ink! metadata/ABI was specified using a version key which
wrapped the whole ink! metadata struct (`MetadataVersioned(InkProject)`).

However, it is nicer to have a proper `version` field as part of the ink! metadata object
so that it can be queried in a less-ambiguous way.

Closes #1289.

### CODE REVIEW

File: crates/metadata/src/lib.rs
Comment (ascjones) [inline crates/metadata/src/lib.rs:72]: I wonder whether we should remove these deprecated version variants entirely. After all they will never be constructed here, and if you want to deserialize legacy versions then you just have to use an older version of the crate.

Comment (HCastano) [reply crates/metadata/src/lib.rs]: Yeah good point, updated

File: crates/metadata/src/lib.rs
Comment (ascjones) [reply crates/metadata/src/lib.rs:72]: Can we just use a manual implementation then for now until 1.62 is propogated everywhere?

**Comment by HCastano** [comment]
The `ink-waterfall` step fails because it's using an old `cargo-contract` `master` instead [of my branch on ](https://github.com/paritytech/cargo-contract/pull/641).

**Comment by HCastano** [comment]
This builds fine with the `staging` Docker image. @cmichi should we use that instead while waiting for the `production` one to be fixed?

**Comment by HCastano** [comment]
@ascjones can you take another look at this?

**Review by sergejparity — APPROVED** [review]

File: crates/metadata/src/lib.rs
Comment (ascjones) [inline crates/metadata/src/lib.rs:73]: Do we still need V3?

**Review by ascjones — APPROVED** [review]
LGTM

File: crates/metadata/src/lib.rs
Comment (HCastano) [reply crates/metadata/src/lib.rs]: Guess not, removed

**Comment by HCastano** [comment]
Let's wait until we sort out the release details of `cargo-contract` before merging this

**Comment by HCastano** [comment]
Gonna :shipit: and get the release train started

**Pull request merged**
